### PR TITLE
Use NoEcho to mask retrieved CF passwords with ****

### PR DESCRIPTION
**Pull request by ianmalott**

### CODE REVIEW

**Comment by mwarkentin** [comment]
Resolves #425

File: api/provider/aws/helpers.go [hunk 1]
```diff
@@ -273,11 +273,16 @@ func (p *AWSProvider) stackUpdate(name string, templateUrl string, changes map[s
...
 		fp, err := formationParameters(string(body))
 
-		// remove params that don't exist in the template
 		for key := range params {
+			// remove params that don't exist in the template
 			if _, ok := fp[key]; !ok {
 				delete(params, key)
 			}
+
+			// exclude Password from params to avoid updating password to "*****"
```
Comment (ddollar) [inline api/provider/aws/helpers.go]: Is there some easy way to catch this higher and eliminate all `NoEcho` parameters from being carried forward into this map in the first place?

**Comment by ddollar** [comment]
Does this affect Console and its creation of Racks in any way?

**Comment by ianmalott** [comment]
@ddollar I got some clarification about Console and rack creation from @mattmanning:

> When console installs a Rack it needs to store the Rack password so it can proxy API calls to it. I would assume David is asking if that will be broken by the Password NoEcho setting.

> I'm guessing that since there's no --password option for the CLI installer that console fetches the password from CloudFormation after the install.

> But would need to confirm.

> If that's the case, we might want to have Console set the Password when it installs so it doesn't have to be fetched back.

Not sure how you'd want to proceed from here, but it seems like this is blocked by some Console and CLI changes.

**Comment by ddollar** [comment]
Would it be possible to test this `UsePreviousValue` logic somehow?

**Comment by ddollar** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;EAYn8PFe&#x2F;421-change-password-parameter-to-noecho-on-rack"> Change Password parameter to NoEcho on Rack</a>

**Comment by ddollar** [comment]
Is this only blocking on tests? If so let's rebase and get it in.

Have we verified that it won't break Console rack installation?

**Comment by ianmalott** [comment]
@ddollar: yep, only blocked on tests (and review of refactoring). Wasn't sure how to test a console-based rack installation, but I tested the following from the CLI:

* `convox install --version <noecho-release>` with `--password` and without
* `convox rack update` to a release with `"NoEcho": "false"` for Password param
* `convox rack params set Password=foobarbaz`

**Comment by mwarkentin** [comment]
@ianmalott @ddollar Might be worth adding a small bit of documentation around what you need to do now if you've lost your password?

**Comment by ianmalott** [comment]
@mwarkentin Thanks for the suggestion. @nzoschke has a doc on the way that describes how to set a new Rack API Key/Password: https://github.com/convox/site/pull/229

**Pull request closed without merging**